### DISCUSSION TITLE
How to launch a "global" log initial function for all rust API?

### DISCUSSION THREAD

**Discussion by dbsxdbsx** [fzyzcjy/flutter_rust_bridge#558] [Q&A] [answered]
I want to use logs, like `debug!`, and `info!` in some of the exposed APIs in rust. And usually, it is necessary to launch an initial function before using them, like this:
```rust
fn main() {
    env_logger::init();  // core

    debug!("this is a debug {}", "message");
    error!("this is printed by default");
}
```
Since there is no `main()` with exposed API, I wonder how to launch `env_logger::init();` here.

**Comment by fzyzcjy** [comment, marked as answer]
Personally I do it in setup functions. http://cjycode.com/flutter_rust_bridge/feature/init.html

**Reply by dbsxdbsx** [reply]
> I do it in setup functions. http://cjycode.com/flutter_rust_bridge/feature/init.html

I tried to mimic `pub fn rust_set_up()` in #486. And I succeeded in revoking `rust_set_up` except that I have to call an empty API to activate the rust module where `rust_set_up` in, before all API can use logger. For example:
```rust
// util_api.rs
pub fn rust_set_up() {
    crate::init_logger_inside("./logs/");
}

pub fn activate_api() {}
```

```dart
// extend_util_api.dart
class UtilApiImplExtend extends UtilApiImpl with FlutterRustBridgeSetupMixin {
  factory UtilApiImplExtend(ffi.DynamicLibrary dylib) =>
      UtilApiImplExtend._raw(UtilApiWire(dylib));

  UtilApiImplExtend._raw(super.inner) : super.raw() {
    setupMixinConstructor();
  }

  @override
  Future<void> setup() async {
    await rustSetUp(hint: FlutterRustBridgeSetupMixin.kHintSetup);
  }
}
```

```dart 
// ffi.dart
import 'dart:ffi';
import 'dart:io' as io;

import 'extend_util_api.dart';
import 'gened_util_api.dart';
import 'gened_rule_api.dart';
import 'gened_ucci_api.dart';

const _base = 'native';

// On MacOS, the dynamic library is not bundled with the binary,
// but rather directly **linked** against the binary.
final _dylib = io.Platform.isWindows ? '$_base.dll' : 'lib$_base.so';

// The late modifier delays initializing the value until it is actually needed,
// leaving precious little time for the program to quickly start up.

final UtilApi utilApi = UtilApiImplExtend(
    io.Platform.isIOS || io.Platform.isMacOS
        ? DynamicLibrary.executable()
        : DynamicLibrary.open(_dylib));
...
```
I have to call `activateApi ` somewhere:
```dart
class HomeController extends GetxController {
  HomeController() {
    utilApi.activateApi();
    ...
}
...
}
```

Now I wonder if the `activateApi` necessary? How to avoid it?

**Reply by fzyzcjy** [reply]
If using that approach seems needed.

Anyway, you may just throw away those "setup" functionality, but instead just create `pub fn my_setup() {}`, and call it like `api.mySetup()` in dart - just treat it as a very normal function.

You may be interested in calling it in, e.g., `main()` of Dart.

**Reply by dbsxdbsx** [reply]
> If using that approach seems needed.
> 
> Anyway, you may just throw away those "setup" functionality, but instead just create `pub fn my_setup() {}`, and call it like `api.mySetup()` in dart - just treat it as a very normal function.
> 
> You may be interested in calling it in, e.g., `main()` of Dart.

Indeed, it is better to put it in `main()`.
